Fix compilation process of built-in Argobots

shintaro-iwasaki opened the pull request:
Previously there are two issues:
1. Built-in Argobots is compiled in `bolt/external/argobots`, though this should only contain source files. It should be built in the build directory (`CMAKE_CURRENT_BINARY_DIR`, e.g., `bolt/build/external/argobots`)
2. Built-in Argobots is installed to `CMAKE_INSTALL_PREFIX` (e.g., `bolt/install`) without the install process (e.g., without `make install`). It should be only compiled in the compilation phase (`make`), and then installed to `CMAKE_INSTALL_PREFIX` in the installation phase (`make install`).

This patch fixes these issues.

These issues were reported by @Meinersbur. Thank you!

Minor points:
1. test before installation: `LIBOMP_ARGOBOTS_INSTALL_DIR` is properly set to the local installation directory, so the test (`llvm-lit`) works well even before installation.
2. `autogen.sh`: Ideally `autogen.sh` should be executed in another directory so that we can keep  `bolt/external/argobots` clean. However, it is not possible with the current `autogen` mechanism. At present, it is executed in `bolt/external/argobots`.
3. `pkgconfig`: Since built-in Argobots is first installed to the local path (`CMAKE_CURRENT_BINARY_DIR`, e.g., `bolt/build/external/argobots`), the `pkgconfig/argobots.pc` points to this directory, not the final installation path (e.g., `CMAKE_INSTALL_PREFIX`). However, LLVM and its subprojects use `llvm-config` instead of `pkgconfig`, so `pkgconfig` is not necessary for built-in Argobots; currently `pkgconfig` is not installed.
However, This seems controversial (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=897275) If LLVM starts to use `pkgconfig`, this problem should be fixed.

Meinersbur commented:
The primary llvm configuration is stored as a [CMake package](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#config-file-packages): Usually ${PREFIX}/lib/cmake/llvm/LLVMConfig.cmake and ${PREFIX}/lib/cmake/clang/ClangConfig.cmake. However, I don't know how to extract its information from outside cmake.

shintaro-iwasaki commented:
@Meinersbur Thanks for the information.  For now the minor issues are derived from `autogen.sh` and `pkgconfig`, which are irrelevant to LLVM and supposedly are not fixed by LLVM's configuration.

However, if you know how to improve my ugly implementation, I am happy to hear your advise! (PR is always welcome)

shintaro-iwasaki commented:
test:jenkins

shintaro-iwasaki commented:
test:jenkins

shintaro-iwasaki commented:
test:jenkins

shintaro-iwasaki committed:
[BOLT] build Argobots in the CMake working directory

Previously, the built-in Argobots was compiled in the Argobots source directory
(e.g., bolt/external/argobots), which messes up the source directory.  This
patch changes the build directory for the built-in Argobots so that it is
configured and build in CMAKE_CURRENT_BINARY_DIR/external/argobots
(e.g., bolt/build/external/argobots).

Reported-by: Michael Kruse <bolt@meinersbur.de>
Signed-off-by: Halim Amer <aamer@anl.gov>

shintaro-iwasaki committed:
[BOLT] temporarily install Argobots in the CMake working directory

Previously, the built-in Argobots was directly installed to the install path
(i.e., CMAKE_INSTALL_PREFIX). This behavior is not favored since actually
installs objects without user's explicit installation (e.g., make install).
This patch changes this behavior. The built-in Argobots is first installed to
the CMake working directory. It is copied to the CMake's installation directory
at installation time.

Because of this change, pkgconfig of the built-in Argobots gets incorrect, while
LLVM and it subprojects should use llvm-config.  This issue should be fixed if
LLVM adopts pkgconfig in the future.

Reported-by: Michael Kruse <bolt@meinersbur.de>
Signed-off-by: Halim Amer <aamer@anl.gov>

shintaro-iwasaki committed:
[BOLT] rename ABT_DIR to ABT_SRC_DIR in CMakeLists.txt

This patch renames ABT_DIR to ABT_SRC_DIR to clarify that this variable points
at the source path, not the build path or installation path.

Reported-by: Michael Kruse <bolt@meinersbur.de>
Signed-off-by: Halim Amer <aamer@anl.gov>
